### ISSUE TITLE
Unpublished Drafts in the main news feed

### ISSUE THREAD

**Issue by WizzardMaker** [ludumdare/ludumdare#1407]
I scrolled through the list, refreshed and saw this.

![image](https://user-images.githubusercontent.com/7768485/33491259-f50ba4c2-d6b9-11e7-92f2-18a2bdb44446.png)

I am logged in, with my account.

**Comment by sgstair** [comment]
Hm, interesting. This looks like a caching error, where your client somehow has a cached copy of the unpublished blog post; In the publishing process though the cache is invalidated correctly, and it is invalidated before the post switches to published status.
I'll have to think more about how this can happen - Either the client somehow managed to cache this node before it was published, or there's something else going on.

**Comment by WizzardMaker** [comment]
It's not even my account though...

**Comment by sgstair** [comment]
Yeah... The post that was having trouble is probably [this one](https://ldjam.com/events/ludum-dare/40/$50977/how-to-joincompo-now-and-what-is-the-game-idea-to-work-ini-dont-see-any-result) (posted immediately after [the other post visible in your image](https://ldjam.com/events/ludum-dare/40/$51462/good-luck-everyone-and-see-ya-in-10-hours))

**Comment by povrazor** [comment]
Did you try reloading the page when you saw that? If so, I think I know what's wrong.

**Comment by sgstair** [comment]
I can think of a very narrow and unlikely race condition that could lead to this condition, but it will resolve itself after 3 minutes (cache timeout). If the node is being requested concurrently with the publish, the invalidate may happen on one thread, and shortly thereafter stale computed data from the other thread could be cached.

**Comment by povrazor** [comment]
I'm suspecting it's not even a race condition. Feeds include a timestamp of the last change. I'm thinking I might have forgotten to compare the modified date of the cached value against what comes with the feed.

**Comment by OfficialExedo** [comment]
Also seeing this as well.

![wat](https://i.gyazo.com/192c26f2bf94b5eb1a142710e124006e.png)

**Comment by sgstair** [comment]
@povrazor when you reload the page, that shouldn't be a concern - all the caches should be gone? I can't really look into it further at the moment but I'll do some deeper digging after LD40

**Comment by povrazor** [comment]
Yeah, I've heard reloading fixes it. I've got some time now, I'll a have a quick look.

**Comment by povrazor** [comment]
Okay. This should be mostly fixed now. Please let me know if you see this behaviour again.

If you do see this again, a note for devs: Switching between the NEWS and FEED tabs didn't trigger a refresh. Possibly because both NEWS and FEED are simple `<ContentTimeline>`'s, and the VDOM->DOM might not be destroying the instance across changes. That said, clicking HOME did work in my tests, as the `<ContentTimeline>`'s DOM element does get destroyed due to its new placement in the greater hierarchy of the page.

Solution would be check out ContentTimeline. There might be some state data that isn't correctly being reset in the constructor.

**Comment by sgstair** [comment]
I did run into this also earlier, though before your comment. I'll review the related changes after LD40

**Comment by sgribushin** [comment]
Same happened to me just now. 
![2c41e1da70ba69bab746135a354102f8](https://user-images.githubusercontent.com/20779250/33525417-70c6ec86-d82f-11e7-93ee-d371c3ed42ba.png)

**Comment by ervion-n** [comment]
![image](https://user-images.githubusercontent.com/14315476/39083977-977850fe-4576-11e8-8697-f0b509ebf714.png)

saw the behaviour again

**Comment by mikekasprzak** [comment]
Being able to reliably reproduce this would help.

**Comment by mikekasprzak** [comment]
Yeah, regrettably I've had almost no time this year to sit down and fix
bugs/finish features (ran out of $$, needed to get a day job, got a game
job which means you work late). Hoping to change that next year once my
schedule lightens up.

When you're seeing a post, it's most definitely a legitimate post. The
problem is some piece of the setup is caching the unpublished post, and
serving that instead of the published one. We're definitely invalidating
cached copies of posts whenever they're updated, but something is either
stomping over the cache, or there's some other mistake somewhere.

Mike Kasprzak
